feat(curriculum): add extra space capturing (#55841)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-regular-expressions-by-building-a-spam-filter/66bb6467b4b7381178420970.md

---
id: 66bb6467b4b7381178420970
title: Step 20
challengeType: 0
dashedName: step-20
---

# --description--

> While this expression does match `1 hundred dollars`, it will not however match `1  hundred  dollars`. Spam messages

Reviewer: ```suggestion
While this expression does match `1 hundred dollars`, it will not however match `1  hundred  dollars`, or `10 dollars`. Spam messages
```

> can and will find a way to bypass flaws in our detector. Let's add a preventive measure and replace the first hard-coded single space with a `\s*` expression and the second with a `\s+`. The `\s` represents a universal space character and the `+` means one 

Reviewer: ```suggestion
can and will find a way to bypass flaws in our detector. Let's improve regex and replace the first literal space with a `\s*` expression and the second with a `\s+`. The `\s` represents a universal space character and the `+` means one 
```

> While this expression does match `1 hundred dollars`, it will not however match `1  hundred  dollars`, or `10 dollars`. Spam messages
> can and will find a way to bypass flaws in our detector. Let's improve regex and replace the first literal space with a `\s*` expression and the second with a `\s+`. The `\s` represents a universal space character and the `+` means one 
> or more of the previous characters. Meanwhile the `*` means zero or more of the following characters. 

Reviewer 2: ```suggestion
While this expression does match `1 hundred dollars`, it will not match `1  hundred  dollars`, or `10 dollars`.

Spam messages can and will find a way to exploit flaws in your detection. Time to improve your regex.

Replace the first literal space with the `\s*` expression. The `\s` character class matches whitespace, such as spaces, tabs, and new lines. The `*` quantifier means "match the previous character 0 or more times".

Replace the second literal space with `\s+`. The `+` quantifier means "match the previous character at least one time".
```

# --hints--

> Your `dollarRegex` must not being using hard-coded spaces. 

Reviewer: ```suggestion
Your `dollarRegex` must not use literal spaces. 
```

```js
assert.notMatch(dollarRegex.source, / /);
```

> Your `dollarRegex` must allow any number of spaces before and after the dollar quantity. 

Reviewer 2: ```suggestion
Your `dollarRegex` must allow any number of spaces before the dollar quantity, and one or more spaces after.
```

---
id: 66bb6467b4b7381178420970
title: Step 20
challengeType: 0
dashedName: step-20
---

# --description--

While this expression does match `1 hundred dollars`, it will not however match `1  hundred  dollars`. Spam messages
can and will find a way to bypass flaws in our detector. Let's add a preventive measure and replace the hard-coded
single spaces with a `\s+` expression. The `\s` represents a universal space character and the `+` means one 

> or more of the previous characters. 

Author: I'm confused. Can you explain why it won't match `10 dollars` after this step? The intended solution is for dollarRegex to be equal to `/[0-9]+s\+(?:hundred|thousand|million|billion)?\s+dollars/i` after this step. 

As for the intent, the point is to keep things mentally clear. A plus sign before an literal empty space can be confusing and obnoxious to deal with. While distracted, it's really easy to get confused and ask "Hey what's that space doing?". Then hit backspace and get stuck on the challenge. Plus when trying to come up with an explanation of how the plus sign works, I'd like there to be a proper symbol to be there. I wanna be able to point at what's being repeated.

> or more of the previous characters. 

Reviewer: Oh yeah, that makes sense for space. It might be good to add more a bit more which characters other than literal space `\s` includes.

> I'm confused. Can you explain why it won't match `10 dollars` after this step? The intended solution is for dollarRegex to be equal to `/[0-9]+\s+(?:hundred|thousand|million|billion)?\s+dollars/i` after this step.

There's `\s+` before and after the optional group. That's one or more spaces twice.

> or more of the previous characters. 

Author: Okay. Now that I can see the issue, what's the best way to resolve the problem?

> or more of the previous characters. 

Reviewer: Not exactly sure, it's easy to make it much more complicated quickly: 
```regex
/[0-9]+\s+(?:(hundred|thousand|million|billion)\s+)?dollars/i
```

What do you think about changing the first `\s+` to `\s*`? 
```regex
/[0-9]+\s*(?:hundred|thousand|million|billion)?\s+dollars/i
```
Matching ie. `2hundred dollars` wouldn't be that bad, would it?

> or more of the previous characters. 

Author: I suppose it wouldn't be so bad.
